CNTRLPLANE-1751: Configurable PKI for OpenShift Internal Certificates

sanchezl opened the pull request:
## Summary

This enhancement introduces the ability to configure cryptographic parameters (key algorithm, key size, and elliptic curves) for certificates and keys generated internally by OpenShift components.

## Motivation

Enterprise customers increasingly face stringent security compliance requirements that mandate specific cryptographic parameters for PKI infrastructure. Currently, OpenShift uses hardcoded defaults (primarily RSA 2048-bit keys) with no mechanism for administrators to adjust these parameters to meet organizational security requirements or compliance mandates.

## Key Features

- New cluster-level `PKI` configuration resource in `config.openshift.io/v1alpha1`
- Support for RSA (2048, 3072, 4096) and ECDSA (P-256, P-384, P-521) algorithms
- Configuration at multiple granularity levels: global defaults, certificate categories (signer, serving, client), and specific named certificates
- Day-1 (installer) support for signer certificate configuration
- Day-2 (operator-driven) support for all certificate types
- Metrics and observability for certificate generation and compliance
- Backward compatible: clusters continue using existing defaults if not configured

## Proposal Details

The enhancement includes:
- New `PKI` CRD in `config.openshift.io/v1alpha1` API group
- `ConfigurablePKI` feature gate for controlled rollout
- Limited Day-1 configuration support via installer for long-lived signer certificates
- Individual operator integration for Day-2 certificate rotation
- Comprehensive validation, metrics, and support procedures

## Tracking

- Jira: https://issues.redhat.com/browse/OCPSTRAT-2271
- Enhancement file: `enhancements/security/internal-pki-config.md`

sanchezl committed:
Add PKI configuration enhancement

sanchezl committed:
Enhance PKI configuration validation using CEL rules for key parameters and certificate overrides. Replace webhooks with in-process validation for improved performance and simplicity.

sanchezl committed:
Refactor PKI key configuration to use RSA/ECDSA-specific subfields and enforce validation rules with unions and CEL.

sanchezl committed:
Refactor PKI configuration to use `CertificateConfig` for improved structure, extensibility, and clarity.

sanchezl committed:
Clarify that the enhancement does not deprecate or remove existing features. Add details about configuring cryptographic parameters while maintaining defaults.

sanchezl committed:
Promote PKI configuration API to `v1`, update compatibility level to GA, and refine graduation criteria for OpenShift 4.21 release

sanchezl committed:
Enhance PKI configuration proposal with improved metric coverage, clarified upgrade scenarios, and refined operational guidance.

sanchezl committed:
Fix lint errors.

sanchezl committed:
Enhance PKI configuration with expanded well-known certificate names and updated validation rules.

sanchezl committed:
Refine PKI configuration XValidation rule by formatting well-known certificate names as a list for improved diffs.

sanchezl committed:
Expand PKI configuration enhancement to include the service-ca-operator's special behavior, future extensibility with dynamic certificate registration, and proposed validation framework.

everettraven reviewed (commented):
Initial pass of the API

sanchezl committed:
Update internal PKI config enhancement with additional reviewers and approvers

sanchezl committed:
Adjust PKI enhancement to use "target OpenShift release" placeholders instead of specific version numbers for flexibility.

sanchezl committed:
Add details on relationships to existing certificate configuration features.

sanchezl commented on enhancements/security/internal-pki-config.md:
````diff
@@ -0,0 +1,1583 @@
+     ```
+   - Compliance checking could be addressed in a future enhancement with a dedicated compliance-checker component if needed
+
+## Open Questions
````
This enhancement only affects OpenShift’s internal certificate generation, not user-provided certificates, trust anchors, TLS parameters, or external CA integration. I have added more details to the enhancement.

sanchezl committed:
clarify how to return to defaults

sanchezl committed:
Add XValidation rule to enforce PKI singleton name constraint.

sanchezl committed:
Make `Key` field in `CertificateConfig` optional.

sanchezl committed:
Remove unused `status` field and subresource annotation from PKI configuration.

sanchezl committed:
use the preferred CEL pattern for KeyConfig discriminated union

everettraven commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1583 @@
+//   'aggregator-client',
+//   'apiserver-proxy-client',
+//   'journal-gatewayd-client',
+// ]",message="certificateName must be a well-known certificate name"
```
Adding to this, I've got no idea whether or not the CRD generators handle a multi-line marker like this. I suspect not.

sanchezl commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1583 @@
+    // No status fields are currently defined. Each certificate-generating operator
+    // independently consumes the PKI configuration and reports status through
+    // its own ClusterOperator status resource.
+}
```
Removed.

sanchezl committed:
Introduce `type` field in PKI configuration. Adjust validation rules and update relevant documentation accordingly.

everettraven commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1695 @@
+
+    // Embed PKIProfile to provide reusable certificate parameter fields.
+    // Go API utilities can use PKIProfile to programmatically define custom profiles.
+    PKIProfile `json:",inline"`
```
For discriminated unions we generally encourage the use of a field whose name matches the discriminator. This allows us to expand in the future to support other values that may require configuration without causing an overlap for the fields that need to be supplied.

I'd expect this to end up looking something like:
```yaml
type: Custom
custom:
  ...
```

sanchezl committed:
Update PKI enhancement to introduce `spec.policy` structure, adjust API fields, and refine validation rules

sanchezl committed:
restructure PKI API and introduce PKICertificateDefinition

- Change spec.policy.type to spec.certificateManagement.mode
- Move custom configuration into spec.certificateManagement.custom field
- Introduce PKICertificateDefinition resource for operators to register their managed certificates
- Add ValidatingAdmissionPolicy for dynamic certificate name validation in overrides

sanchezl committed:
simplify PKI API: remove PKICertificateDefinition and shift to category-based overrides

- Eliminate `PKICertificateDefinition` resource and dynamic certificate registration
- Replace `overrides` with explicit category fields (`signerCertificates`, `servingCertificates`, `clientCertificates`)
- Refine `KeyConfig` and enforce stricter validation rules

patrickdillon reviewed (commented):
Well done. This looks great from the installer perspective.

sanchezl commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1296 @@
+- ECDSA: Faster handshakes, less CPU intensive
+
+**Recommendations:**
+- Use larger RSA keys (3072/4096) for long-lived signer certificates where generation is rare
```
@patrickdillon Thanks for the review. 

> _The added details were very helpful, but leads me to another question: are APIServer.servingCerts and this PKI API mutually exclusive? In which case we might need some sort of validation to prevent them from both being set._

There is a section about the [relationship to existing configuration features](https://github.com/sanchezl/enhancements/blob/8b54be1e2588fd28ba4e85ec0358996f6117744c/enhancements/security/internal-pki-config.md#relationship-to-existing-certificate-configuration-features).

TLDR: They're complementary — no mutual exclusion or validation is needed because they configure different aspects. `servingCerts` specifies which certificates to use for named hosts, while `PKI` configures how certificates are generated (algorithm, key size). Both can coexist.

> _Why not always use ECDSA? Are there still reasons to prefer RSA?_

We might. Once we have enough CI coverage to analyze performance impacts and surface any compatibility issues we can revisit our defaults. For the initial implementation, we're keeping the existing defaults, but possibly bumping the signer certs to (RSA 4096).

everettraven commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1296 @@
+
+The compatibility level is enforced through the `+openshift:compatibility-gen:level` annotation and will be validated by the API review process. The annotation will change from `level=4` to `level=1` when the API is promoted to v1.
+
+#### PKI Resource
```
If it helps, you could probably just point to your now approved openshift/api PR here.

everettraven reviewed (commented):
This enhancement is approved from an API review perspective.

sanchezl committed:
replace inline API definitions with reference and concise key aspects

sanchezl committed:
add OKE compatibility statement

sanchezl committed:
update PKI enhancement authors list

sanchezl committed:
refine PKI configuration: update certificate naming conventions and clarify recovery processes

sanchezl committed:
expand PKI configuration: add new certificate signers and serving certificates, update naming conventions

sanchezl committed:
add cluster-monitoring-operator certificate names

Register the monitoring operator's GRPC TLS certificate names in the
well-known certificates registry:
- monitoring.grpc-tls-signer (CA for Thanos GRPC TLS)
- monitoring.grpc-tls-serving (Prometheus GRPC server)
- monitoring.grpc-tls-client (Thanos querier client)

sanchezl commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1583 @@
+    // defaults specifies the default certificate configuration
+    // for all certificates unless overridden by category or specific
+    // certificate configuration.
+    // If not specified, uses platform defaults (typically RSA 2048).
```
A small number of leaf certificates already use ECDSA P-256 today, but signed with an RSA 2048 signer. `Default` mode will default to P-256 for leaf certificates, keeping these intact.

benluddy commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1207 @@
+
+The `certificateManagement.mode` field controls how operators determine cryptographic parameters:
+
+- **`Unmanaged`**: Operators use their existing hardcoded defaults. This is the initial mode set during upgrade to ensure zero behavior change. Most operators default to RSA 2048, though a small number of leaf certificates already use ECDSA P-256 today (kubelet client/serving certs generated via the Kubernetes CSR mechanism, and OLM-managed certs).
```
Is there going to be a process to phase out Unmanaged?

benluddy commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1207 @@
+
+1. **Category Override**: If the corresponding category field (`signerCertificates`, `servingCertificates`, or `clientCertificates`) is set, use those parameters
+2. **Default Configuration**: Use the `defaults` configuration
+3. **Hardcoded Defaults**: Use hardcoded defaults (typically RSA 2048)
```
Aren't hardcoded defaults only used in Unmanaged mode?

benluddy commented on enhancements/security/internal-pki-config.md:
```diff
@@ -0,0 +1,1207 @@
+  - **Defaults** configuration if specified in the PKI resource
+  - Hardcoded defaults (RSA 2048) when mode is `Unmanaged`
+
+**Important limitation**: Individual service serving certificates cannot be configured independently because:
+- They are generated dynamically in response to service annotations
+- There are potentially thousands of services in a cluster
+- Certificate names are service-specific (e.g., `<namespace>/<service-name>`) and not well-known at the platform level
```
This should be mentioned in the documentation, since it means an admin needs be aware that changing this configuration can break non-platform workloads that get their serving certs from service-ca-operator.

oceanc80 reviewed (commented):
/lgtm
